scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala

```diff
@@ -224,12 +202,42 @@ private[scio] case class WrappedBCoder[T](u: BCoder[T]) extends BCoder[T] {
     u.registerByteSizeObserver(value, observer)
 }
 
-private object WrappedBCoder {
+private[scio] object WrappedBCoder {
   def create[T](u: BCoder[T]): BCoder[T] =
     u match {
       case WrappedBCoder(_) => u
       case _                => new WrappedBCoder(u)
     }
+
+  def appendMaterializationStack[X <: Throwable](
+    cause: X,
+    additionalMessage: Option[String],
+    baseStack: Array[StackTraceElement]
+  ): X = {
+    val existingStack = cause.getStackTrace
+    val messageItem =
+      additionalMessage.map(msg => new StackTraceElement("Due to " + msg, "", "", 0)).toArray
+
+    val adjustedStack = messageItem ++ existingStack ++ baseStack
+    cause.setStackTrace(adjustedStack)
+    cause
+  }
+
+  def prepareStackTrace: Array[StackTraceElement] =
+    Array(
+      new StackTraceElement(
```

Comment:
The parameters are class, method, file name and line number. This is a little hacky? How does it look in an actual stacktrace? I wonder if there're better ways to cascade these, like caused by or suppressed?

Reply:
Yes, the call to StackTraceElement does look a bit ugly (hence the "dash-scissor-dash" thingy). Let's see if we can do better.

I don't have a screenshot at the moment, since the whole point was to let an EOFException unchanged (and it ends up being caught by the Flink runtime as expected), but the prototype looked clear enough ☺ I'll whip up something ASAP for demonstration.

The problem with the other options:

* to generate a a"Caused By:", I would need to wrap the exception into another exception… Which is the problem in the first place. There might have been a way to wrap into an exception of the same type using reflection, if we could rely on the knowledge that every exception type must have a ctor that takes a Throwable or a String and a Throwable, but there is no such rule in the Java spec. 

* I don't know that we can say "suppressed" here. In the specific case that led me here, the EOFException ends up being suppressed, indeed. Let's say (hypothetically) that the coder must throw an exception such as "NoReallyICannotSerializeThisException", that the Runtime is _not_ aware of. In that hypothetical case, we probably (i) don't want to suppress (current behaviour: `WrappedException(NoReallyICannotSerializeThisException(original stack), materialization stack)` surfaces into a Job Failed, with this PR: `NoReallyICannotSerializeThisException(original stack / "—✂—" / materialization stack)` surfaces into the JobFailed. And _maybe_ the client is the owner of the Coder[T] involved and knows what to do with it.

Reply:
(at the very minimum, I'd name the parameters and put an explanatory comment)